extensions/opc/src/opc.cpp

```diff
@@ -0,0 +1,461 @@
+    UA_Client_disconnect(client);
+  }
+  UA_Client_delete(client);
+}
+
+void setCertificates(ClientPtr& clientPtr, const std::vector<char>& certBuffer, const std::vector<char>& keyBuffer) {
+  UA_ClientConfig *cc = UA_Client_getConfig(clientPtr.get());
+  cc->securityMode = UA_MESSAGESECURITYMODE_SIGNANDENCRYPT;
+
+  UA_ByteString certByteString = UA_BYTESTRING_ALLOC(certBuffer.data());
+  UA_ByteString keyByteString = UA_BYTESTRING_ALLOC(keyBuffer.data());
+
+  UA_ClientConfig_setDefaultEncryption(cc, certByteString, keyByteString,
+                                       nullptr, 0,
+                                       nullptr, 0);
+
+  UA_ByteString_delete(&certByteString);
+  UA_ByteString_delete(&keyByteString);
+}
+
+ClientPtr connect(const std::string& url, const std::shared_ptr<core::logging::Logger>& logger, const std::string& username, const std::string& password) {
+  UA_Client *client = UA_Client_new();
+  UA_ClientConfig_setDefault(UA_Client_getConfig(client));
+  UA_StatusCode retval;
+  if(username.empty()) {
+    retval = UA_Client_connect(client, url.c_str());
+  } else {
+    retval = UA_Client_connect_username(client, "opc.tcp://localhost:53530/OPCUA/SimulationServer",
+        username.c_str(), password.c_str());
+  }
+  if (retval != UA_STATUSCODE_GOOD) {
+    logger->log_error("Failed to connect to %s (%s)", url.c_str(), UA_StatusCode_name(retval));
+    UA_Client_delete(client);
+    return ClientPtr(nullptr, &disconnect);
+  }
+  logger->log_info("Successfully connected to %s", url.c_str());
+  return ClientPtr(client, &disconnect);
+}
+
+bool isConnected(const ClientPtr &ptr) {
+  UA_Client* client = ptr.get();
+  if(!client ) {
+    return false;
+  }
+  return UA_Client_getState(client) != UA_CLIENTSTATE_DISCONNECTED;
+}
+
+void traverse(ClientPtr& clientPtr, UA_NodeId nodeId, std::function<nodeFoundCallBackFunc> cb, const std::string& basePath) {
+
+  UA_Client *client = clientPtr.get();
+  UA_BrowseRequest bReq;
+  UA_BrowseRequest_init(&bReq);
+  bReq.requestedMaxReferencesPerNode = 0;
+  bReq.nodesToBrowse = UA_BrowseDescription_new();
+  bReq.nodesToBrowseSize = 1;
+
+  UA_NodeId_copy(&nodeId, &bReq.nodesToBrowse[0].nodeId);
+  bReq.nodesToBrowse[0].resultMask = UA_BROWSERESULTMASK_ALL;
+
+  UA_BrowseResponse bResp = UA_Client_Service_browse(client, bReq);
+
+  UA_BrowseRequest_deleteMembers(&bReq);
+
+  for(size_t i = 0; i < bResp.resultsSize; ++i) {
+    for(size_t j = 0; j < bResp.results[i].referencesSize; ++j) {
+      UA_ReferenceDescription *ref = &(bResp.results[i].references[j]);
+      if (cb(clientPtr, ref, basePath)) {
+        if (ref->nodeClass == UA_NODECLASS_VARIABLE || ref->nodeClass == UA_NODECLASS_OBJECT) {
+          std::string browsename((char *) ref->browseName.name.data, ref->browseName.name.length);
+          traverse(clientPtr, ref->nodeId.nodeId, cb, basePath + browsename);
+        }
+      } else {
+        UA_BrowseResponse_deleteMembers(&bResp);
```

Comment:
As this is repeated a few lines down, using a util::ScopeGuard would be less error-prone.

Reply:
Good idea, added!